apps/site/assets/ts/helpers/service.ts

```diff
@@ -1,155 +1,197 @@
+  service: Service,
+  currentDate: Date = new Date()
+): boolean => {
+  if (hasNoRating(service)) {
+    return false;
   }
-
-  return [
-    {
-      type: "future",
-      servicePeriod: `starts ${formattedDate(startDateObject)}`,
-      service
-    }
-  ];
+  const ratingStartDate = new Date(service.rating_start_date!);
+  const ratingEndDate = new Date(service.rating_end_date!);
+  return currentDate <= ratingStartDate && currentDate <= ratingEndDate;
 };
 
-export const serviceDate = (
-  service: Service | ServiceWithServiceDate
-): string => {
-  const { start_date: startDate, end_date: endDate } = service;
-  if (service.start_date === service.end_date) {
-    return holidayDate(service);
-  }
+export const groupServicesByDateRating = (
+  services: Service[],
+  currentDate: Date = new Date()
+): Dictionary<Service[]> =>
+  _.groupBy(services, (service: Service) => {
+    if (service.typicality === "holiday_service") {
+      return "Holiday Schedules";
+    }
 
-  const startDateObject = new Date(startDate);
-  const endDateObject = new Date(endDate);
-  if ("service_date" in service) {
-    // Get unix timestamps
-    const serviceDateTime = new Date(service.service_date).getTime();
-    const startDateTime = startDateObject.getTime();
-    const endDateTime = endDateObject.getTime();
+    // if there's no rating end date, this is probably a CR or Ferry.
+    // don't use the rating in that case.
+    if (hasNoRating(service)) {
+      if (isInCurrentService(service, currentDate)) {
+        return "Current Schedules";
+      }
+      if (isInFutureService(service, currentDate)) {
+        return "Future Schedules";
+      }
+    } else {
+      const ratingStartDate = new Date(service.rating_start_date!);
+      const ratingEndDate = new Date(service.rating_end_date!);
+
+      if (
+        isInCurrentRating(service, currentDate) &&
+        service.typicality === "typical_service"
+      ) {
+        return `Current Schedules (${
+          service.rating_description
+        }, ends ${shortDate(ratingEndDate)})`;
+      }
 
-    if (serviceDateTime >= startDateTime && serviceDateTime <= endDateTime) {
-      return `ends ${formattedDate(endDateObject)}`;
+      if (isInFutureRating(service, currentDate)) {
+        return `Future Schedules (${
+          service.rating_description
+        }, starts ${shortDate(ratingStartDate)})`;
+      }
     }
 
-    if (serviceDateTime < startDateTime) {
-      return `starts ${formattedDate(startDateObject)}`;
-    }
-  }
-  return startToEnd(startDateObject, endDateObject);
+    return "Other Schedules";
+  });
+
+// sort the option groups
+const optGroupOrder: { [key: string]: number } = {
+  Current: 4,
+  Holiday: 3,
+  Future: 2,
+  Other: 1
 };
 
-type MonthInteger = number | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11;
+export const optGroupComparator = (groupA: string, groupB: string): number => {
+  const a = groupA.split(" ")[0];
+  const b = groupB.split(" ")[0];
```

Comment:
This seems fragile — is there any way we can accomplish the same goal without relying on the string names of the option groups?

Reply:
I'm not sure how to address this since the whole point of this function is to sort the string names of the option groups?

Reply:
I think it's sort of the same thing you addressed by adding the `enum` — if the user-facing names of the option groups changed, they would sort in a different order (despite that our intention is e.g. "the one that represents current schedules should be first" and not "the one that starts with the string `Current` should be first"). It'd be great if `optGroupOrder` could now use the `enum`, since we have that.